machinelearning/tensorflow.md

```diff
@@ -0,0 +1,349 @@
+
+```
+# To enable XLA auto clustering, and to generate .dot files
+XLA_FLAGS="--xla_dump_to=/tmp/generated  --xla_dump_hlo_as_dot" TF_XLA_FLAGS="--tf_xla_auto_jit=2 --tf_xla_cpu_global_jit" python run.py --backend=tf --scenario=SingleStream
+
+# To convert to .svg for visual graph
+sudo apt install graphviz
+dot -Tsvg <.dot file to be converted> >  <output .svg name>
+
+eg:
+dot -Tsvg 1645569101166784.module_0000.cluster_51__XlaCompiledKernel_true__XlaHasReferenceVars_false__XlaNumConstantArgs_0__XlaNumResourceArgs_0_.80.before_optimizations.dot > module0_before_opts.svg
+
+dot -Tsvg 1645569101166784.module_0000.cluster_51__XlaCompiledKernel_true__XlaHasReferenceVars_false__XlaNumConstantArgs_0__XlaNumResourceArgs_0_.80.cpu_after_optimizations.dot > module0_after_opts.svg
+
+```
+
+2. Once the XLA graph looks as expected (without too mnay duplicated nodes), check how the ops are emitted. Currently XLA framework logging is under the same TF CPP logging, and level 1 is sufficient to get most of the info traces.
+
+```
+# Enable TF CPP framework logging
+TF_CPP_MAX_VLOG_LEVEL=1
+
+```
+Then look for the emitter level traces to understand how each op for a given shape is lowered to LLVM IR. For ACL supported shapes, the below traces are expected for dot and conv IR emitters.
+
+```
+# ACL runtime traces
+__xla_cpu_runtime_ACLBatchMatMulF32
+
+__xla_cpu_runtime_ACLConv2DF32
+
+# Eigen runtime traces
+__xla_cpu_runtime_EigenBatchMatMulF32
+
+__xla_cpu_runtime_EigenConv2DF32
+
+```
+If the shapes are emitted by eigen runtime, check the source build configuration to make sure 'mkl_aarch64' bazel config is enabled (which internally enables '--define=build_with_acl=true' bazel configuration). If the LLVM IR still emits Eigen runtime, please raise an ssue on [ACL github](https://github.com/ARM-software/ComputeLibrary) with the operator and shape details to understand what constraints are limiting the shape to get accelerated with ACL gemm kernels.
+
+3. The above triaging steps cover the performance issues due to the missing compile time, runtime or library level optimizations across the Tensorflow stack. If the troubleshooting is inconclusive or stuck with any step mentioned above, please raise a support request on [aws-graviton-getting-started](https://github.com/aws/aws-graviton-getting-started) github.
+
+
+# Enable Tensorflow serving with onednn+acl backend
```

Comment:
should be much higher?

Reply:
since this is again dependent on source build, i put it after the source build section, may be it can be before XLA.

Reply:
Could you put the instructions to enable tensorflow serving with the defaults at the top but then reference this howto section for incorporating the optimized software packages?

Reply:
ok, sounds good.

Reply:
btw, I realized that there are no dockers/wheels available for serving, the PR I mentioned in the below section adds aarch64 build  and docker options to the repo. However, we have our AWS DLC built with the serving API interface, I will provide them here.